Python/pystate.c

```diff
@@ -782,6 +781,13 @@ PyThreadState_Clear(PyThreadState *tstate)
     Py_CLEAR(tstate->async_gen_finalizer);
 
     Py_CLEAR(tstate->context);
+
+    if (tstate->on_delete != NULL) {
+        // This will unblock any joining threads.
+        tstate->on_delete(tstate->on_delete_data);
+        tstate->on_delete = NULL;
+        tstate->on_delete_data = NULL;
```

Comment:
This change looks to be unrelated. If it's related, I would prefer to see it as a separated PR to prepare that one.

Same comment for other thread changes:

```
    // All threading module threads are marked as "done" later
    // in PyThreadState_Clear().
```

and

```
        // This will unblock any joining threads.
        // We also do this in PyThreadState_Clear(), but do it here to be sure.
```